### ISSUE TITLE
Replace gotype

### ISSUE THREAD

**Issue by maticmeznar** [alecthomas/gometalinter#91]
According to Google, `x/tools/cmd/gotype` is on life-support only. Replacement is described in  https://github.com/golang/go/issues/12703.

**Comment by alecthomas** [comment]
Well, that's interesting but not really practical at this stage. Two issues:

1. It wouldn't work on Go 1.5.
2. It's not go-gettable.

**Comment by djui** [comment]
:+1:

**Comment by alecthomas** [comment]
There doesn't seem to be any intention to make gotype a separate command, so my guess is I'll have to remove it from gometalinter. This isn't as bad as it sounds, as quite a few linters use the types package and report the same errors. That said, it's kind of annoying that an existing useful command has become orphaned.

**Comment by alecthomas** [comment]
Actually, I'm reopening this because some errors (most obviously, misspellings of keywords) don't get reported by any other linter. It would be excellent if anyone interested in this could comment in golang/go#12703, as I will be.

**Comment by alecthomas** [comment]
Filed a new request as golang/go#14727

**Comment by bbuck** [comment]
Misspelling of keywords seems like you'd catch that with syntax highlight/syntax errors? I've just recently stumbled upon gometalinter and have been using standard tools (excluding go type) before this and have never thought that having a tool catch misspellings of keywords to be useful, or even thought it necessary. Some of the other included linters I've briefly thought about -- such as dead code and unused consts/struct fields definitely but this one just seems a little odd to desire to keep around (at least for that one particular reason).

**Comment by client9** [comment]
@bbuck It sounded strange to me too, but some IDEs use it in place of a full compile which might be slow.  see  golang/go#12703 and golang/go#14727 for details.

**Comment by alecthomas** [comment]
That is just one example, there are probably more instances. Also if you misspell "range" as "rnge" in code you want an error, not a spell check.

**Comment by akutz** [comment]
I'm also having to use [this workaround](https://github.com/joefitzgerald/gometalinter-linter/issues/24#issuecomment-198952334) as `gotype` fails to ever find the vendored testify import for my test packages as they don't get compiled to normal archives in `$GOPATH/pkg`. Note if I nuke `$GOPATH/pkg/$GOOS_$GOARCH/PATH_TO_MY_PROJ` the `gotype` tool fails to import *anything*. Apparently `gotype` only works on compiled archives for resolving vendored imports. The order seems to be:

1. Check import path via source file
2. If it doesn't exist then check the import path via the compiled archive package
3. If no such archive exists then return an error

:(

```
[0]akutz@pax:libstorage$ env GO15VENDOREXPERIMENT=1 gotype -a .
libstorage_test.go:18:2: could not import github.com/stretchr/testify/assert (can't find import: github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert)
libstorage_test.go:563:2: undeclared name: assert
libstorage_test.go:564:2: undeclared name: assert
libstorage_test.go:566:2: undeclared name: assert
libstorage_test.go:568:2: undeclared name: assert
libstorage_test.go:570:2: undeclared name: assert
libstorage_test.go:571:2: undeclared name: assert
libstorage_test.go:572:2: undeclared name: assert
libstorage_test.go:573:2: undeclared name: assert
```

However, running tests work just fine :)

```
[0]akutz@pax:libstorage$ go test .
ok  	github.com/emccode/libstorage	8.216s
[0]akutz@pax:libstorage$ 
```

Thus the issue is with the `gotype` tool, not the source since `go test` works just fine.

**Comment by akutz** [comment]
Ha! Apparently I was [dead on](https://github.com/golang/go/issues/12703#issuecomment-144429128) with my guess about the way `gotype` works:

> The gotype tool assumes that you have recently done a "go install" of all dependency packages that are imported by the source files that it type checks, since it parses the pkg.a object code files produced by the compiler in order to obtain type information. (This limitation is unfortunate and not well advertised.)

**Comment by akutz** [comment]
Okay, I've discovered another hack-ish solution that appears necessary as disabling `gotype` from the metalinter basically breaks any meaningful feedback on any sources at all. 

The following command `go test -i` will install all of the dependencies of the tests, but not run the tests. And the dependencies are installed using the vendored sources.

Here's an example where I execute `go test -i -x -a` to build the test dependencies while showing what's actually happening under the covers:

```
$ go test -i -x -a
...
mkdir -p $WORK/github.com/emccode/libstorage/vendor/github.com/stretchr/testify/
cd /Users/akutz/Projects/go/src/github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert
/Users/akutz/.go/1.6/pkg/tool/darwin_amd64/compile -o $WORK/github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert.a -trimpath $WORK -p github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert -complete -buildid ecd74dfaa40e22b707d732f5f77dc5f0c99f71b0 -importmap github.com/davecgh/go-spew/spew=github.com/emccode/libstorage/vendor/github.com/stretchr/testify/vendor/github.com/davecgh/go-spew/spew -importmap github.com/pmezard/go-difflib/difflib=github.com/emccode/libstorage/vendor/github.com/stretchr/testify/vendor/github.com/pmezard/go-difflib/difflib -D _/Users/akutz/Projects/go/src/github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert -I $WORK -I /Users/akutz/Projects/go/pkg/darwin_amd64 -pack ./assertion_forward.go ./assertions.go ./doc.go ./errors.go ./forward_assertions.go ./http_assertions.go
...
```

As seen above, the vendored version of testify *is* being installed. Here's proof:

### Remove All Compiled Packages
This command removes all the packages compiled for my project as well as the packages and sources for the `testify` project.
```
$ rm -fr $GOPATH/pkg/*/github.com/emccode/libstorage $GOPATH/src/github.com/stretchr $GOPATH/pkg/*/github.com/stretchr
```

### Execute Gotype w Massive Fail
This `gotype` execution fails due to no packages being present to resolve vendored dependencies.
```
$ gotype -a -v .
libstorage_tests_test.go
libstorage.go
libstorage_test.go
libstorage_tests_test.go:8:2: could not import github.com/emccode/libstorage/api/client (can't find import: github.com/emccode/libstorage/api/client)
libstorage.go:39:2: could not import github.com/akutz/gofig (can't find import: github.com/emccode/libstorage/vendor/github.com/akutz/gofig)
libstorage.go:41:2: could not import github.com/emccode/libstorage/api/client (can't find import: github.com/emccode/libstorage/api/client)
libstorage.go:42:2: could not import github.com/emccode/libstorage/api/registry (can't find import: github.com/emccode/libstorage/api/registry)
libstorage.go:43:2: could not import github.com/emccode/libstorage/api/server (can't find import: github.com/emccode/libstorage/api/server)
libstorage.go:44:2: could not import github.com/emccode/libstorage/api/types/context (can't find import: github.com/emccode/libstorage/api/types/context)
libstorage.go:45:2: could not import github.com/emccode/libstorage/api/types/drivers (can't find import: github.com/emccode/libstorage/api/types/drivers)
libstorage_test.go:14:6: could not import github.com/Sirupsen/logrus (can't find import: github.com/emccode/libstorage/vendor/github.com/Sirupsen/logrus)
libstorage_test.go:15:2: could not import github.com/akutz/gofig (can't find import: github.com/emccode/libstorage/vendor/github.com/akutz/gofig)
libstorage_test.go:16:2: could not import github.com/akutz/goof (can't find import: github.com/emccode/libstorage/vendor/github.com/akutz/goof)
```

### Install Project
Installing my project's non-tests will cause `gotype` to work for my project's non-tests.
```
$ go install .
```

### Execute Gotype w Test Fail
This time `gotype` will fail only for the tests.
```
$ gotype -a -v .
libstorage_tests_test.go
libstorage_test.go
libstorage.go
libstorage_test.go:18:2: could not import github.com/stretchr/testify/assert (can't find import: github.com/emccode/libstorage/vendor/github.com/stretchr/testify/assert)
libstorage_test.go:563:2: undeclared name: assert
libstorage_test.go:564:2: undeclared name: assert
libstorage_test.go:566:2: undeclared name: assert
libstorage_test.go:568:2: undeclared name: assert
libstorage_test.go:570:2: undeclared name: assert
libstorage_test.go:571:2: undeclared name: assert
libstorage_test.go:572:2: undeclared name: assert
libstorage_test.go:573:2: undeclared name: assert
```

### Install Test Dependencies
Using the `-i` flag we can install the test dependencies. They do not get installed in `$GOPATH/pkg/$GOOS_$GOARCH` but rather `
```
[2]akutz@pax:libstorage$ go test -i 
```

### Test Dependencies Location
The installed test dependency packages are not in `$GOPATH/pkg/$GOOS_$GOARCH` but rather in `$GOPATH/pkg/$GOOS_$GOARCH/github.com/emccode/libstorage/vendored`.
```
$ ls $GOPATH/pkg/*/github.com/emccode/libstorage/vendor
total 0
drwxr-xr-x+  5 akutz  staff   170B Mar 20 18:42 ./
drwxr-xr-x+  4 akutz  staff   136B Mar 20 18:42 ../
drwxr-xr-x+ 14 akutz  staff   476B Mar 20 18:43 github.com/
drwxr-xr-x+  3 akutz  staff   102B Mar 20 18:42 golang.org/
drwxr-xr-x+  4 akutz  staff   136B Mar 20 18:42 gopkg.in/
```

### Execute Gotype Successfully
Finally `gotype` works on tests as well.
```
[0]akutz@pax:libstorage$ gotype -a -v .
libstorage_tests_test.go
libstorage.go
libstorage_test.go
43.839456ms (3 files, 901 lines, 20552 lines/s)
[0]akutz@pax:libstorage$ 
```

**Comment by alecthomas** [comment]
There's a lot of discussion in #105, but I think basically we agreed that

1. gotype is not a good tool for linting (or anything?)
2. It's not really gometalinter's responsibility to fix issues in other linters.

**Comment by akutz** [comment]
Agreed.

Follow-up, what's your suggestion for a replacement to `gotype`? It sounded like you had a few in mind, and I wouldn't mind supplying a patch to incorporate one or more of them.

**Comment by joefitzgerald** [comment]
I think it's important to note that `gocode` also works on `.a` files and there's a valid argument to be made that `go install ./...` should be something periodically run by the user or automatically (e.g. on save) by the user's editor, if they so choose.

**Comment by joefitzgerald** [comment]
(Incidentally the #1 most common issue people had with `go-plus `3.x` was related to gocode, not having up to date suggestions, and the need to either run `go install ./...` periodically or run `gocode set autobuild true` (which @nsf [:wave:] himself recommends _against_).

**Comment by akutz** [comment]
Hi @maticmeznar, don't listen to @joefitzgerald. He's nobody. Just the inventor of go-plus, that's all :)

Also, @joefitzgerald, thanks. You just answered the question as to why autocomplete doesn't work half the time :( I knew it used gocode, but I didn't realize it relied on the packages.

(or why autocomplete is out of date)

**Comment by akutz** [comment]
By the way, many thanks to both of you, @alecthomas & @joefitzgerald, for your great projects. I particuarly dislike you though @alecthomas because `gometalinter` replaces much of the [tools work](https://github.com/akutz/gomk/tree/master/.gomk/include/tools/source) I thought was valuable as part of my Gomk project :) Now I can just replace it all with `gometalinter` :)

**Comment by joefitzgerald** [comment]
Just to circle back on this and close the loop:

* `go install .` is now run on save
* `gotype` is no longer enabled by default (although you can turn it back on and it will function; it'll just give you duplicate error messages )

This has appeared to resolve the issues being experienced by go-plus whether they use gotype or not.

**Comment by bcomnes** [comment]
Ran into this bug the other day via metaliner: https://github.com/golang/go/issues/14918  

Looks like it lacks support for vendored deps and so I have to turn it off :sad_trombone:

**Comment by GarethLewin** [comment]
gotype 100% for sure supports vendored packages.

You need to make sure your vendored packages are installed as mentioned in this thread, make sure to `go install` or `go build -i`

**Comment by joefitzgerald** [comment]
Does gotype report anything that go install would not?

**Comment by bcomnes** [comment]
@GarethLewin Ahh! Interesting.  I'm going to have to read about what go install actually does.  Looks like it builds out `$GOPATH/pkg` including vendor paths.  Thanks!

@joefitzgerald I vendor'd some code and tried `go run` and `go build`, both worked fine, but gotype still was complaining about not being able to import the vendor code.

**Comment by joefitzgerald** [comment]
Disable gotype for now. The default args now do this.

**Comment by alecthomas** [comment]
@akutz I don't think a better solution exists, unfortunately. `go install` might be the best alternative, but having a linter arbitrarily installing packages and binaries doesn't seem like a great solution :(

**Comment by alecthomas** [comment]
@joefitzgerald out of curiosity, how does go-plus handle completion via gocode (which I believe also requires libraries be installed)?

**Comment by joefitzgerald** [comment]
@alecthomas in prior versions, (<= 3.x) the #1 issue with go-plus was people complaining about out of date autocomplete suggestions. gocode works on .a files, so the solution was to either run `go install` periodically, or run `gocode set autobuild true` (which functionally does a similar thing, and @nsf recommends against autobuild).

I had quite a bit of code that was dedicated to running `go bulld` and redirecting the output to a temporary directly where the output would be thrown away, so I was basically already doing 99% of the work required for an install anyway, on save. After the `gotype` misstep, I built [builder-go](https://github.com/joefitzgerald/builder-go) which resolves the issue by running `go install .` on save.

**Comment by joefitzgerald** [comment]
By the way, @alecthomas, you should hop on https://gophers.slack.com every now and then :smile:.

**Comment by mcandre** [comment]
Is gotype disabled in gometalinter HEAD? gotype reports tons of spurious warnings about "undeclared name" for perfectly functioning imports.

**Comment by GarethLewin** [comment]
mcandre are you sure to install your pkgs before running gotype?
